### ISSUE TITLE
[AutoDiff] Incorrect tangent vector for generic function returning optional

### ISSUE THREAD

**Issue by BradLarson** [apple/swift#64223]
**Description**
A differentiable generic function that returns an Optional yields an incorrect tangent vector. 

**Steps to reproduce**

Place the following in a file:

```swift
import _Differentiation

@differentiable(reverse)
func givesWrongTangentVector<Element>(x: Element) -> Element? where Element: Differentiable {
    return x
}
@differentiable(reverse)
func f(x: Double) -> Double {
    let y = givesWrongTangentVector(x: x)
    return y!
}
@differentiable(reverse)
func g(x: Double) -> Double {
    return x
}
print(valueWithGradient(at: 0.0, of: f))
print(valueWithGradient(at: 0.0, of: g))
```

and build and run it.

**Expected behavior**

If the `givesWrongTangentVector` function was differentiable, you'd expect the results to be 

```
(value: 0.0, gradient: 1.0)
(value: 0.0, gradient: 1.0)
```

but they instead are

```
(value: 0.0, gradient: 0.0)
(value: 0.0, gradient: 1.0)
```

However, it's unclear whether `givesWrongTangentVector` should be differentiable. A concrete version of the same function:

```swift
@differentiable(reverse)
func givesWrongTangentVector(x: Double) -> Double? {
    return x
}
```

is marked as not being differentiable by the compiler and won't build.

**Environment**
- Swift compiler version info: all swift.org toolchains, this is not a recent regression

**Comment by asl** [comment]
@asavonic Fixed by https://github.com/apple/swift/pull/68300?

**Comment by BradLarson** [comment]
@asl @asavonic Looks like it is indeed fixed by #68300, gradients are now correct in local testing.
